src/core/server/services/users/users.ts

```diff
@@ -658,21 +692,133 @@ export async function updateUsernameByID(
 export async function updateRole(
   mongo: Db,
   tenant: Tenant,
-  user: Pick<User, "id">,
+  user: Pick<User, "id" | "role" | "moderationScopes">,
   userID: string,
   role: GQLUSER_ROLE
 ) {
   if (user.id === userID) {
     throw new Error("cannot update your own user role");
   }
 
-  return updateUserRole(mongo, tenant.id, userID, role);
+  // only admins can promote to admin
+  if (user.role !== GQLUSER_ROLE.ADMIN && role === GQLUSER_ROLE.ADMIN) {
+    throw new Error("only admins can update users to administrators");
+  }
+
+  // you can only update users if you're currently an admin or moderator
+  if (![GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR].includes(user.role)) {
+    throw new Error("you have insufficient permissions to promote users");
+  }
+
+  // make sure that site moderators cannot upgrade users to
+  // organization moderators
+  const siteModeratorsEnabled = hasFeatureFlag(
+    tenant,
+    GQLFEATURE_FLAG.SITE_MODERATOR
+  );
+  if (
+    siteModeratorsEnabled &&
+    user.role === GQLUSER_ROLE.MODERATOR &&
+    isModerationScoped(user.moderationScopes)
+  ) {
+    throw new Error(
+      "scoped moderators cannot promote users to organization level moderators"
+    );
+  }
+
+  return await updateUserRole(mongo, tenant.id, userID, role);
+}
+
+export async function promoteUser(
+  mongo: Db,
+  tenant: Tenant,
+  user: Pick<User, "id" | "role" | "moderationScopes">,
```

Comment:
I'd suggest calling this `viewer` to disassociate this from the user we're promoting a bit further.
```suggestion
  viewer: Pick<User, "id" | "role" | "moderationScopes">,
```